src/System.ComponentModel.TypeConverter/src/System/ComponentModel/VersionConverter.cs

```diff
@@ -0,0 +1,87 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Globalization;
+
+namespace System.ComponentModel
+{
+    /// <summary>
+    ///    <para>Provides a type converter to convert Version objects to and
+    ///       from various other representations.</para>
+    /// </summary>
+    public class VersionConverter : TypeConverter
+    {
+        /// <summary>
+        ///    <para>Gets a value indicating whether this converter can convert an object in the
+        ///       given source type to a Version.</para>
+        /// </summary>
+        public override bool CanConvertFrom(ITypeDescriptorContext context, Type sourceType)
+        {
+            if (sourceType == null)
+                throw new ArgumentNullException(nameof(sourceType));
+
+            return sourceType == typeof(string) || sourceType == typeof(Version);
+        }
+
+        /// <summary>
+        ///    <para>Gets a value indicating whether this converter can
+        ///       convert an object to the given destination type using the context.</para>
+        /// </summary>
+        public override bool CanConvertTo(ITypeDescriptorContext context, Type destinationType)
+        {
+            return destinationType == typeof(string) || destinationType == typeof(Version);
+        }
+
+        /// <summary>
+        ///    <para>Converts the given object to a Version.</para>
+        /// </summary>
+        public override object ConvertFrom(ITypeDescriptorContext context, CultureInfo culture, object value)
+        {
+            if (value is string versionString)
+            {
+                // Let the Version constructor throw any informative exceptions
+                return new Version(versionString);
+            }
+
```

Comment:
The above might throw an ArgumentException.  Is that desired?   Should it instead be a FormatException?  What do the docs for ConvertFrom state?

Reply:
This was discussed in https://github.com/dotnet/corefx/issues/28594#issuecomment-377323389 and @safern agreed on
> let's keep it simple and just let the constructor handle it itself

But the doc from [ConvertFrom](https://docs.microsoft.com/en-us/dotnet/api/system.componentmodel.typeconverter.convertfrom?view=netcore-2.1#System_ComponentModel_TypeConverter_ConvertFrom_System_ComponentModel_ITypeDescriptorContext_System_Globalization_CultureInfo_System_Object_) says
> Exceptions
> [NotSupportedException](https://docs.microsoft.com/en-us/dotnet/api/system.notsupportedexception?view=netcore-2.1)
> The conversion cannot be performed.

And `NotSupportedException` is used in the base `TypeConverter` class when the value is not of a supported type (`string` or `Version`).

I'm not sure what’s the best solution here.

Reply:
@safern, I'll leave it up to you.  Just seems strange to me that we'd differ from what's doc'd here.  Are the docs wrong / are the other type converters already inconsistent and throw exceptions like ArgumentException?

cc: @danmosemsft

Reply:
I talked to @stephentoub offline since I missed this discussion, I think it is a fair point to try to keep consistency across newly added converters and try to fit what is best. As a conclusion after looking at other converters this are the points I would like to mention:

- The docs are referring to the base class as @0xced states which for me is correct as that should be thrown only when the type converting from is not a supported type to convert from. In this case not `string` or `Version` which we're doing in all other converters. However it might be worth mentioning that in overridden methods we could throw `ArgumentException` or `FormatException` 

- The other exceptions thrown are `ArgumentException` and `FormatException` in all other converters, but they are not leaving it to the Parser to throw the exception, they are catching the exception and throwing a custom one. 

So I think the right thing to do here is to catch the exception and throw a `FormatException` as I suggested earlier in the issue. The reason why I believe that is the right outcome is because, that way the exception is clear that when trying to convert from a supported type we couldn't convert it because the format of the value was incorrect. That is what for me is the clearest approach and that way we would maintain a "sort of consistency" where we only throw ArgumentException, FormatException or NotSupportedException. 

Version.Parse throws this exceptions

ArgumentNullException | input is null.
-- | --
ArgumentException | input has fewer than two or more than four version components.
ArgumentOutOfRangeException | At least one component in input is less than zero.
FormatException | At least one component in input is not an integer.
OverflowException | At least one component in input represents a number that is greater than Int32.MaxValue.

So I wouldn't mind catching everything but FormatException and converting that to ArgumentException if we think that would be a best fit.

Reply:
I agree that throwing `FormatException` is better. Unfortunately, I only looked at `GuidConverter` and `UriTypeConverter` which both don't catch any exceptions thrown by the constructors.

It is now addressed in 004ef11348398f7dbdd65405d6fe1a9e65889076 and rebased on master.